iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+    if (top_item->node->meta.expiry_time < curr_time) {
+      this->erase(top_item->node->meta.key);
+      expiry_queue.pop();
+    } else { // if the first item isn't expired-- the rest won't be either (queue is sorted)
+      return false;
+    }
+  }
+  return true;
+}
+
+template <class C>
+const size_t
+RefCountCachePartition<C>::count()
+{
+  return this->items;
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::copy(Vec<RefCountCacheHashEntry *> &items)
+{
+  for (RefCountCachePartition<C>::iterator_type i = this->item_map.begin(); i != this->item_map.end(); ++i) {
+    RefCountCacheHashEntry *val = refCountCacheHashingValueAllocator.alloc();
+    val->set(i.m_value->item.get(), i.m_value->meta.key, i.m_value->meta.size, i.m_value->meta.expiry_time);
+    items.push_back(val);
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::metric_inc(RefCountCache_Stats metric_enum, int64_t data)
+{
+  if (this->rsb) {
+    RecIncrGlobalRawStatCount(this->rsb, metric_enum, data);
+  }
+}
+
+template <class C>
+TSHashTable<RefCountCacheHashing> *
+RefCountCachePartition<C>::get_map()
+{
+  return &this->item_map;
+}
+
+// The header for the cache, this is used to check if the serialized cache is compatible
+// The implementation of this class must be in here, as this is used by template
+// classes, and c++ rocks that way
+class RefCountCacheHeader
+{
+public:
+  unsigned int magic;
+  VersionNumber version;
+  VersionNumber object_version; // version passed in of whatever it is we are caching
+
+  RefCountCacheHeader(VersionNumber object_version = VersionNumber())
+    : magic(REFCOUNTCACHE_MAGIC_NUMBER), object_version(object_version)
+  {
+    this->version.ink_major = REFCOUNTCACHE_MAJOR_VERSION;
+    this->version.ink_minor = REFCOUNTCACHE_MINOR_VERSION;
+  };
+
+  bool
+  operator==(const RefCountCacheHeader other) const
+  {
+    return (this->magic == other.magic && this->version.ink_major == other.version.ink_major &&
+            this->version.ink_minor == other.version.ink_minor &&
+            this->object_version.ink_major == other.object_version.ink_major &&
+            this->object_version.ink_minor == other.object_version.ink_minor);
+  }
+
+  bool
+  compatible(RefCountCacheHeader *other)
+  {
+    return (this->magic == other->magic && this->version.ink_major == other->version.ink_major &&
+            this->object_version.ink_major == other->version.ink_major);
+  };
+};
+
+// This continuation is responsible for persisting RefCountCache to disk
+// To avoid locking the partitions for a long time we'll do the following per-partition:
+//    - lock
+//    - copy ptrs (bump refcount)
+//    - unlock
+//    - persist
+//    - remove ptrs (drop refcount)
+// This way we only have to hold the lock on the partition for the time it takes to get Ptr<>s to all items in the partition
+template <class C> class RefCountCacheSync : public Continuation
```

Comment:
Classes should be named with nouns because they generate concrete objects. How about ``RefCountCacheSerializer``?